test/framework/machine_helpers.go

```diff
@@ -21,32 +21,106 @@ import (
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
-	"github.com/pkg/errors"
 
+	"github.com/pkg/errors"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
 )
 
+type GetMachinesByClusterInput struct {
+	Lister      Lister
+	ClusterName string
+	Namespace   string
+}
+
+type GetControlPlaneMachinesByClusterInput struct {
+	GetMachinesByClusterInput
+}
+
+// GetControlPlaneMachinesByCluster returns the control plane Machine objects for a cluster.
+// Important! this method relies on labels that are created by the CAPI controllers during the first reconciliation, so
+// it is necessary to ensure this is already happened before calling it.
+func GetControlPlaneMachinesByCluster(ctx context.Context, input GetControlPlaneMachinesByClusterInput) []clusterv1.Machine {
+	options := append(byClusterOptions(input.ClusterName, input.Namespace), controlPlaneMachineOptions()...)
+
+	machineList := &clusterv1.MachineList{}
+	Expect(input.Lister.List(ctx, machineList, options...)).To(Succeed(), "Failed to list MachineList object for Cluster %s/%s", input.Namespace, input.ClusterName)
+
+	return machineList.Items
+}
+
+type GetMDMachinesByClusterInput struct {
+	GetMachinesByClusterInput
+	Deployment clusterv1.MachineDeployment
+}
+
+// GetMDMachinesByCluster returns Machine objects for a cluster belonging to a machine deployment.
+// Important! this method relies on labels that are created by the CAPI controllers during the first reconciliation, so
+// it is necessary to ensure this is already happened before calling it.
+func GetMDMachinesByCluster(ctx context.Context, input GetMDMachinesByClusterInput) []clusterv1.Machine {
+	opts := byClusterOptions(input.ClusterName, input.Namespace)
+	opts = append(opts, machineDeploymentOptions(input.Deployment)...)
+
+	machineList := &clusterv1.MachineList{}
+	Expect(input.Lister.List(ctx, machineList, opts...)).To(Succeed(), "Failed to list MachineList object for Cluster %s/%s", input.Namespace, input.ClusterName)
+
+	return machineList.Items
+}
+
 // WaitForMachinesToBeUpgradedInput is the input for WaitForMachinesToBeUpgraded.
 type WaitForMachinesToBeUpgradedInput struct {
```

Comment:
I think we should have 1 input structure for each method, no matter if they are the same.
This will make future change less painful (given the compatibility rules)